From 6153bb87878196b01a04cb832837633c4e354ec9 Mon Sep 17 00:00:00 2001
From: xnnnsets <muhamadsolihin019501@gmail.com>
Date: Sat, 2 Aug 2025 20:16:31 +0700
Subject: [PATCH] net: Don't reduce TTL/HL Some mobile carriers compare TTL/HL
 values to restrict tethering. Just skip the decrement to make it harder for
 them.

Signed-off-by: Alexander Winkowski <dereference23@outlook.com>
---
 net/ipv4/ip_forward.c | 11 ++++++-----
 net/ipv6/ip6_output.c |  8 ++++----
 2 files changed, 10 insertions(+), 9 deletions(-)

diff --git a/net/ipv4/ip_forward.c b/net/ipv4/ip_forward.c
index 29730edda220..8eeb8bee5ac6 100644
--- a/net/ipv4/ip_forward.c
+++ b/net/ipv4/ip_forward.c
@@ -109,7 +109,7 @@ int ip_forward(struct sk_buff *skb)
 
 	skb_forward_csum(skb);
 	net = dev_net(skb->dev);
-
+#if 0
 	/*
 	 *	According to the RFC, we must first decrease the TTL field. If
 	 *	that reaches zero, we must reply an ICMP control message telling
@@ -117,7 +117,7 @@ int ip_forward(struct sk_buff *skb)
 	 */
 	if (ip_hdr(skb)->ttl <= 1)
 		goto too_many_hops;
-
+#endif
 	if (!xfrm4_route_forward(skb))
 		goto drop;
 
@@ -139,10 +139,10 @@ int ip_forward(struct sk_buff *skb)
 	if (skb_cow(skb, LL_RESERVED_SPACE(rt->dst.dev)+rt->dst.header_len))
 		goto drop;
 	iph = ip_hdr(skb);
-
+#if 0
 	/* Decrease ttl after skb cow done */
 	ip_decrease_ttl(iph);
-
+#endif
 	/*
 	 *	We now generate an ICMP HOST REDIRECT giving the route
 	 *	we calculated.
@@ -164,11 +164,12 @@ int ip_forward(struct sk_buff *skb)
 	 */
 	 icmp_send(skb, ICMP_DEST_UNREACH, ICMP_SR_FAILED, 0);
 	 goto drop;
-
+#if 0
 too_many_hops:
 	/* Tell the sender its packet died... */
 	__IP_INC_STATS(net, IPSTATS_MIB_INHDRERRORS);
 	icmp_send(skb, ICMP_TIME_EXCEEDED, ICMP_EXC_TTL, 0);
+#endif
 drop:
 	kfree_skb(skb);
 	return NET_RX_DROP;
diff --git a/net/ipv6/ip6_output.c b/net/ipv6/ip6_output.c
index 3c3a08b3ec9f..24c7e4bbe3ec 100644
--- a/net/ipv6/ip6_output.c
+++ b/net/ipv6/ip6_output.c
@@ -524,7 +524,7 @@ int ip6_forward(struct sk_buff *skb)
 		if (ip6_call_ra_chain(skb, ntohs(opt->ra)))
 			return 0;
 	}
-
+#if 0
 	/*
 	 *	check and decrement ttl
 	 */
@@ -535,7 +535,7 @@ int ip6_forward(struct sk_buff *skb)
 		kfree_skb(skb);
 		return -ETIMEDOUT;
 	}
-
+#endif
 	/* XXX: idev->cnf.proxy_ndp? */
 	if (net->ipv6.devconf_all->proxy_ndp &&
 	    pneigh_lookup(&nd_tbl, net, &hdr->daddr, skb->dev, 0)) {
@@ -622,9 +622,9 @@ int ip6_forward(struct sk_buff *skb)
 	hdr = ipv6_hdr(skb);
 
 	/* Mangling hops number delayed to point after skb COW */
-
+#if 0
 	hdr->hop_limit--;
-
+#endif
 	return NF_HOOK(NFPROTO_IPV6, NF_INET_FORWARD,
 		       net, NULL, skb, skb->dev, dst->dev,
 		       ip6_forward_finish);
-- 
2.43.0

